tree/ntuple/v7/src/RNTupleJoinTable.cxx

```diff
@@ -78,51 +61,115 @@ void ROOT::Experimental::Internal::RNTupleJoinTable::Build(RPageSource &pageSour
 {
-   EnsureBuilt();
-
    if (valuePtrs.size() != fJoinFieldNames.size())
       throw RException(R__FAIL("number of value pointers must match number of join fields"));
 
-   std::vector<NTupleJoinValue_t> joinFieldValues;
-   joinFieldValues.reserve(valuePtrs.size());
+   std::vector<JoinValue_t> castJoinValues;
+   castJoinValues.reserve(valuePtrs.size());
 
    for (unsigned i = 0; i < valuePtrs.size(); ++i) {
-      joinFieldValues.push_back(CastValuePtr(valuePtrs[i], fJoinFieldValueSizes[i]));
+      castJoinValues.push_back(CastValuePtr(valuePtrs[i], fJoinFieldValueSizes[i]));
+   }
+
+   if (const auto &entries = fMapping.find(RCombinedJoinFieldValue(castJoinValues)); entries != fMapping.end()) {
+      return &entries->second;
    }
 
-   auto entryIdxs = fJoinTable.find(RCombinedJoinFieldValue(joinFieldValues));
+   return nullptr;
+}
+
+//------------------------------------------------------------------------------
+
+std::unique_ptr<ROOT::Experimental::Internal::RNTupleJoinTable>
+ROOT::Experimental::Internal::RNTupleJoinTable::Create(const std::vector<std::string> &fieldNames)
+{
+   return std::unique_ptr<RNTupleJoinTable>(new RNTupleJoinTable(fieldNames));
+}
+
+ROOT::Experimental::Internal::RNTupleJoinTable &
+ROOT::Experimental::Internal::RNTupleJoinTable::Add(RPageSource &pageSource, PartitionKey_t partitionKey)
+{
+   auto joinMapping = std::unique_ptr<REntryMapping>(new REntryMapping(pageSource, fJoinFieldNames));
+   fPartitions[partitionKey].emplace_back(std::move(joinMapping));
+
+   return *this;
+}
 
-   if (entryIdxs == fJoinTable.end())
+std::vector<ROOT::NTupleSize_t>
+ROOT::Experimental::Internal::RNTupleJoinTable::GetEntryIndexes(const std::vector<void *> &valuePtrs,
+                                                                PartitionKey_t partitionKey) const
+{
+   auto partition = fPartitions.find(partitionKey);
+   if (partition == fPartitions.end())
       return {};
 
-   return entryIdxs->second;
+   std::vector<ROOT::NTupleSize_t> entryIdxs{};
+
+   for (const auto &joinMapping : partition->second) {
+      auto entriesForMapping = joinMapping->GetEntryIndexes(valuePtrs);
+      if (entriesForMapping)
+         entryIdxs.insert(entryIdxs.end(), entriesForMapping->begin(), entriesForMapping->end());
+   }
+
+   return entryIdxs;
+}
+
+std::unordered_map<ROOT::Experimental::Internal::RNTupleJoinTable::PartitionKey_t, std::vector<ROOT::NTupleSize_t>>
+ROOT::Experimental::Internal::RNTupleJoinTable::GetPartitionedEntryIndexes(
+   const std::vector<void *> &valuePtrs, const std::vector<PartitionKey_t> &partitionKeys) const
+{
+   std::unordered_map<PartitionKey_t, std::vector<ROOT::NTupleSize_t>> entryIdxs{};
+
+   for (const auto &partitionKey : partitionKeys) {
+      auto entriesForPartition = GetEntryIndexes(valuePtrs, partitionKey);
+      entryIdxs[partitionKey].insert(entryIdxs[partitionKey].end(), entriesForPartition.begin(),
+                                     entriesForPartition.end());
```

Comment:
Should this check
```suggestion
      auto entriesForPartition = GetEntryIndexes(valuePtrs, partitionKey);
      if (!entriesForPartition.empty())
         entryIdxs[partitionKey].insert(entryIdxs[partitionKey].end(), entriesForPartition.begin(),
                                        entriesForPartition.end());
```
? The question is what do we want to return if there is no index in a particular partition: an empty `vector` or omit the key from the map. The other overload seems to do the latter, and that's also what I would semantically expect. (we should probably add a test for this overload, if I'm not mistaken)

Reply:
Oh yeah, thanks for catching this! I think for now indeed it's fine to use the latter behavior and let time and future developments tell whether whether this was the right intuition. I've expanded the tests as well.